grovel/grovel.lisp

```diff
@@ -179,68 +195,140 @@ int main(int argc, char**argv) {
+       :do (alexandria:copy-file src-file dest-file))
+    (format nil "-I~A" (truename local-include-dir))))
+
+
+;; This function is here to maintain backwards compatibility and is used by
+;; tests.
+(defun process-grovel-file (input-file &optional (output-defaults input-file))
+  (let* ((c-file (make-c-file-name output-defaults "__grovel"))
+         (exe-file (make-exe-file-name c-file))
+         (lisp-file (tmp-lisp-file-name c-file)))
+    (process-grovel-file* input-file lisp-file c-file exe-file nil)))
 
 ;;; *PACKAGE* is rebound so that the IN-PACKAGE form can set it during
 ;;; *the extent of a given grovel file.
-(defun process-grovel-file (input-file &optional (output-defaults input-file))
+(defun process-grovel-file* (input-file dest-lisp-file c-file exe-file absolute-cache-dir)
   (with-standard-io-syntax
-    (let* ((c-file (generate-c-file input-file output-defaults))
-           (exe-file (make-exe-file-name c-file))
-           (lisp-file (tmp-lisp-file-name c-file))
-           (inputs (list (cc-include-grovel-argument) c-file)))
-      (handler-case
-          (link-executable exe-file inputs)
-        (error (e)
-          (grovel-error "~a" e)))
-      (invoke exe-file lisp-file)
-      lisp-file)))
+    (multiple-value-bind (forms feature-expressions)
+        (read-grovel-file input-file)
+      (let* ((cached-lisp-file (feature-specific-cache-file
+                                dest-lisp-file absolute-cache-dir
+                                feature-expressions))
+             (*local-includes* nil))
+        (if (and cached-lisp-file (file-exists-p cached-lisp-file))
+            (process-grovel-file-from-cache dest-lisp-file c-file
+                                            cached-lisp-file)
+            (process-grovel-file-from-scratch forms dest-lisp-file c-file
+                                              exe-file cached-lisp-file))
+        dest-lisp-file))))
+
+(defun process-grovel-file-from-cache (dest-lisp-file c-file cached-lisp-file)
+  (touch-file c-file)
+  (alexandria:copy-file cached-lisp-file dest-lisp-file
+                        :if-to-exists :supersede))
+
+(defun process-grovel-file-from-scratch (forms dest-lisp-file c-file exe-file
+                                         cached-lisp-file)
+  (generate-c-file c-file forms)
+  (let* ((lisp-file (tmp-lisp-file-name c-file))
+         (inputs (list (cc-include-grovel-argument)
+                       c-file)))
+    ;;
+    (when *local-includes*
+      (push (copy-local-includes-to-cache c-file) inputs))
+    ;;
+    (handler-case (link-executable exe-file inputs)
+      (error (e) (grovel-error "~a" e)))
+    (invoke exe-file lisp-file)
+    (rename-file-overwriting-target lisp-file dest-lisp-file)
+    (when cached-lisp-file
+      (ensure-directories-exist
+       (pathname-directory-pathname cached-lisp-file))
+      (alexandria:copy-file dest-lisp-file cached-lisp-file
+                            :if-to-exists :supersede))))
```

Comment:
Here, as in the old version, we do the bulk of the work. 
- We copy the `local-include` files to the designated directory
- We link, invoke and move the file to it's final location
- We then, if the user has specified a `cache-dir` cache the result.